decidim-core/lib/decidim/command.rb

```diff
@@ -40,5 +41,23 @@ def method_missing(method_name, ...)
     def respond_to_missing?(method_name, include_private = false)
       @caller.respond_to?(method_name, include_private)
     end
+
+    def with_events(with_transaction: false, before: true, after: true, &block)
```

Comment:
1. We could extract it to their own concern, but i would not do it now, as there are no visible use cases for me at the moment. I am proposing you to leave it as it is now and refactor at some later date when we really need to use it in some other place. 

2. I would like to have the before and after keywords, since, i will dispatching this event also for "create actions", and most of the times we pass the resource as event payload, we may enter in a realm where the resource is being called even though does not exist. 

I do not support having `ActiveSupport::Notification.publish(...)` directly, as this is the root cause of this entire PR. Based on your comment from here. https://github.com/decidim/decidim/pull/10151#discussion_r1129066881 where we need to replace `dispatch_system_event` with the proposed method. Also, the same method is being implemented also in: create command which you commented to apply the same principle: 
![image](https://github.com/decidim/decidim/assets/105683/214fe86a-73bf-4f1c-938c-d44bd01feb10)

You could see a more elaborate example in:  #11117

Reply:
> 1. We could extract it to their own concern, but i would not do it now, as there are no visible use cases for me at the moment. I am proposing you to leave it as it is now and refactor at some later date when we really need to use it in some other place.

OK, fine for me.

> 2. I would like to have the before and after keywords, since, i will dispatching this event also for "create actions", and most of the times we pass the resource as event payload, we may enter in a realm where the resource is being called even though does not exist.
> 
> I do not support having `ActiveSupport::Notification.publish(...)` directly, as this is the root cause of this entire PR. Based on your comment from here. [#10151 (comment)](https://github.com/decidim/decidim/pull/10151#discussion_r1129066881) where we need to replace `dispatch_system_event` with the proposed method. 

Yes, I agree that not in normal situations. I was just thinking why do we need to call the method at all if we are only using one of those.

I checked the example you posted but what is the problem there if the resource is `nil` during the before hook?

Reply:
Regarding combining the `with_events` method with the transaction option, as @alecslupu explained it yesterday, this is to ensure the correct ordering of publishing the events and then starting the transaction block. The events need to be published/dispatched outside of the transaction which is why it makes sense to wrap into this same method to try to prevent bugs when dealing with events and transactions. There has been situations before where the events didn't work correctly because they were published during the transaction block.